src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -2856,61 +2876,46 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
-            // for every declaration.
-            tmp = nullptr;
-        } else {
-            ASR::expr_t *value = ASRUtils::EXPR(tmp);
             ASR::ttype_t *target_type = ASRUtils::expr_type(target);
-            ASR::ttype_t *value_type = ASRUtils::expr_type(value);
+            ASR::ttype_t *value_type = ASRUtils::expr_type(tmp_value);
             if( ASR::is_a<ASR::Pointer_t>(*target_type) &&
                 ASR::is_a<ASR::Var_t>(*target) ) {
-                if( !ASR::is_a<ASR::GetPointer_t>(*value) ) {
+                if( !ASR::is_a<ASR::GetPointer_t>(*tmp_value) ) {
                     throw SemanticError("A pointer variable can only "
                                         "be associated with the output "
                                         "of pointer() call.",
-                                        value->base.loc);
+                                        tmp_value->base.loc);
                 }
                 if( !ASRUtils::check_equal_type(target_type, value_type) ) {
                     throw SemanticError("Casting not supported for different pointer types. Received "
                                         "target pointer type, " + ASRUtils::type_to_str_python(target_type) +
                                         " and value pointer type, " + ASRUtils::type_to_str_python(value_type),
                                         x.base.base.loc);
                 }
-                tmp = ASR::make_Assignment_t(al, x.base.base.loc, target, value, nullptr);
-                return ;
+                tmp = ASR::make_Assignment_t(al, x.base.base.loc, target, tmp_value, nullptr);
+                continue;
             }
-
-            cast_helper(target, value, true);
-            value_type = ASRUtils::expr_type(value);
+            cast_helper(target, tmp_value, true);
+            value_type = ASRUtils::expr_type(tmp_value);
             if (!ASRUtils::check_equal_type(target_type, value_type)) {
                 std::string ltype = ASRUtils::type_to_str_python(target_type);
                 std::string rtype = ASRUtils::type_to_str_python(value_type);
                 diag.add(diag::Diagnostic(
                     "Type mismatch in assignment, the types must be compatible",
                     diag::Level::Error, diag::Stage::Semantic, {
                         diag::Label("type mismatch ('" + ltype + "' and '" + rtype + "')",
-                                {target->base.loc, value->base.loc})
+                                {target->base.loc, tmp_value->base.loc})
                     })
                 );
                 throw SemanticAbort();
             }
             ASR::stmt_t *overloaded=nullptr;
-            tmp = ASR::make_Assignment_t(al, x.base.base.loc, target, value,
+            tmp = ASR::make_Assignment_t(al, x.base.base.loc, target, tmp_value,
                                     overloaded);
+            tmp_vec.push_back(tmp);
```

Comment:
Rather than assigning to both tmp and tmp_vec, why not use the convention that if it is just one, it will be in tmp, and if `tmp == nullptr`, then it will be in `tmp_vec`?

Reply:
Here is how I am thinking about it:

* tmp is not null: we have exactly one statement, the most common case
* tmp is null: the tmp_vec is used to communicate what is returned: tmp_vec.size() == 0 (nothing), size() == 2 or more (two or more statements).

We could in principle only use tmp_vec even for 1 statement (and remove `tmp`), but I think the `tmp` convention is used for expressions as well, and it seems like the above idea is better.

Reply:
Yes. I agree.

Reply:
In this case we should update other code that currently returns `tmp=null` to mean "none" to ensure it sets `tmp_vec.size() == 0`.

Reply:
Yeah. Otherwise incorrect set of statements mind end up getting into ASR. Well we should always clear the `tmp_vec` after pushing all the elements from it to the body. That way we will know that its always empty once used.